File: tests/fixtures/gen_fixtures.R
# Fixture generation script for clintrials tests
#
# This script generates test fixtures for the Bayesian CRM tests.
# It uses the `bcrm` R package to calculate the expected posterior
# probabilities of dose-limiting toxicity (DLT) and the next
# recommended dose for a set of predefined clinical trial scenarios.
#
# The output of this script is a set of CSV files that are used by
# the `pytest` suite to validate the clintrials Python implementation.
#
# To run this script, you need to have R installed with the `bcrm`
# and `mvtnorm` packages.
#
# To install dependencies (in R):
# > install.packages(c("bcrm", "mvtnorm"))
#
# To run the script (in bash):
# > Rscript tests/fixtures/gen_fixtures.R

library(bcrm)

# Scenario 1: Standard case
# Define prior probabilities of toxicity
p_tox_prior_1 <- c(0.1, 0.2, 0.3, 0.4)
# Define the target toxicity level
target_tox_1 <- 0.3
# Define the observed data: dose levels and DLTs
doses_1 <- c(1, 1, 2, 2, 3, 3)
dlt_1 <- c(0, 0, 0, 1, 1, 1)
data_1 <- data.frame(
  patient = 1:length(doses_1),
  dose = doses_1,
  tox = dlt_1
)

# Run the bcrm analysis
# Using method="exact" to match the Python implementation's approach
crm_1 <- bcrm(
  data = data_1,
  p.tox0 = p_tox_prior_1,
  target.tox = target_tox_1,
  ff = "logit1",
  prior.alpha = list(3, 0, 1.34^2),
  constrain = FALSE,
  sdose.calculate = "mean",
  pointest = "mean",
  method = "exact",
  stop = list(nmax = length(doses_1))
)

# Extract the posterior probabilities of DLT and the next recommended dose
posterior_dlt_probs_1 <- crm_1$ndose[[length(crm_1$ndose)]]$est
next_dose_1 <- crm_1$ndose[[length(crm_1$ndose)]]$ndose

# Scenario 2: No toxicities observed
p_tox_prior_2 <- c(0.05, 0.1, 0.2, 0.35, 0.5)
target_tox_2 <- 0.2
doses_2 <- c(1, 1, 1, 2, 2, 2)
dlt_2 <- c(0, 0, 0, 0, 0, 0)
data_2 <- data.frame(
  patient = 1:length(doses_2),
  dose = doses_2,
  tox = dlt_2
)

crm_2 <- bcrm(
  data = data_2,
  p.tox0 = p_tox_prior_2,
  target.tox = target_tox_2,
  ff = "logit1",
  prior.alpha = list(3, 0, 1.34^2),
  constrain = FALSE,
  sdose.calculate = "mean",
  pointest = "mean",
  method = "exact",
  stop = list(nmax = length(doses_2))
)

posterior_dlt_probs_2 <- crm_2$ndose[[length(crm_2$ndose)]]$est
next_dose_2 <- crm_2$ndose[[length(crm_2$ndose)]]$ndose

# Combine the results into data frames
posterior_dlt_probs_df <- rbind(
  data.frame(scenario = 1, dose = 1:length(posterior_dlt_probs_1), prob = posterior_dlt_probs_1),
  data.frame(scenario = 2, dose = 1:length(posterior_dlt_probs_2), prob = posterior_dlt_probs_2)
)

next_dose_df <- data.frame(
  scenario = c(1, 2),
  next_dose = c(next_dose_1, next_dose_2)
)

# Write the data frames to CSV files
write.csv(posterior_dlt_probs_df, "tests/fixtures/expected_posterior_dlt_probs.csv", row.names = FALSE)
write.csv(next_dose_df, "tests/fixtures/next_dose_recommendations.csv", row.names = FALSE)

print("Fixtures generated successfully.")

File: tests/fixtures/gen_fixtures.py
import os
import numpy as np
import pandas as pd
from scipy.stats import norm

from clintrials.core.math import inverse_logit1, logit1
from clintrials.dosefinding.crm import crm

def generate_fixtures():
    # Scenario 1
    p_tox_prior_1 = [0.1, 0.2, 0.3, 0.4]
    target_tox_1 = 0.3
    doses_1 = [1, 1, 2, 2, 3, 3]
    dlt_1 = [0, 0, 0, 1, 1, 1]

    _, _, _, prob_tox_1 = crm(
        prior=p_tox_prior_1,
        target=target_tox_1,
        toxicities=dlt_1,
        dose_levels=doses_1,
        F_func=logit1,
        inverse_F=inverse_logit1,
        beta_dist=norm(loc=0, scale=np.sqrt(1.34)),
        method="bayes",
        plugin_mean=False,
    )
    recommended_dose_1 = np.argmin(np.abs(np.array(prob_tox_1) - target_tox_1)) + 1

    # Scenario 2
    p_tox_prior_2 = [0.05, 0.1, 0.2, 0.35, 0.5]
    target_tox_2 = 0.2
    doses_2 = [1, 1, 1, 2, 2, 2]
    dlt_2 = [0, 0, 0, 0, 0, 0]

    _, _, _, prob_tox_2 = crm(
        prior=p_tox_prior_2,
        target=target_tox_2,
        toxicities=dlt_2,
        dose_levels=doses_2,
        F_func=logit1,
        inverse_F=inverse_logit1,
        beta_dist=norm(loc=0, scale=np.sqrt(1.34)),
        method="bayes",
        plugin_mean=False,
    )
    recommended_dose_2 = np.argmin(np.abs(np.array(prob_tox_2) - target_tox_2)) + 1

    # Combine the results into data frames
    posterior_dlt_probs_df = pd.DataFrame({
        "scenario": [1] * len(prob_tox_1) + [2] * len(prob_tox_2),
        "dose": list(range(1, len(prob_tox_1) + 1)) + list(range(1, len(prob_tox_2) + 1)),
        "prob": list(prob_tox_1) + list(prob_tox_2)
    })

    next_dose_df = pd.DataFrame({
        "scenario": [1, 2],
        "next_dose": [recommended_dose_1, recommended_dose_2]
    })

    # Write the data frames to CSV files
    out_probs = os.path.join("tests", "fixtures", "expected_posterior_dlt_probs.csv")
    out_next = os.path.join("tests", "fixtures", "next_dose_recommendations.csv")
    
    os.makedirs(os.path.dirname(out_probs), exist_ok=True)
    
    posterior_dlt_probs_df.to_csv(out_probs, index=False)
    next_dose_df.to_csv(out_next, index=False)
    
    print("Fixtures generated successfully.")  # noqa: T201

if __name__ == "__main__":
    generate_fixtures()
